Adding support of path-dependent GADT reasoning

Linyxus opened the pull request:
Previously, GADT reasoning only works on type parameters. This PR supports GADT reasoning for path-dependent types (that is, type members addressed from stable paths).

It mainly makes the following two things possible:

1. It allows us to utilize the GADT constraints of type members implied by the pattern matching. Example:

```Scala
  type typed[E <: Expr, V] = E & { type T = V }

  trait Expr { type T }
  case class LitInt(x: Int) extends Expr { type T = Int }
  case class Add(e1: Expr typed Int, e2: Expr typed Int) extends Expr { type T = Int }
  case class LitBool(x: Boolean) extends Expr { type T = Boolean }
  case class Or(e1: Expr typed Boolean, e2: Expr typed Boolean) extends Expr { type T = Boolean }

  def eval(e: Expr): e.T = e match
    case LitInt(x) => x  // for example: here we can derive the bound e.T = Int
    case Add(e1, e2) => eval(e1) + eval(e2)
    case LitBool(b) => b
    case Or(e1, e2) => eval(e1) || eval(e2)
```

2. It enables the compiler to record GADT bounds for constrainable path-dependent types even if their path is not pattern-matched against. Example:
```Scala
trait Expr[+X]
case class LitInt(x: Int) extends Expr[Int]

trait TypeTag { type A }

def foo(p: TypeTag, e: Expr[p.A]): p.A = e match
  case LitInt(x) => x  // here we derive the bounds p.A >: Int, just like what we do for type parameters
```

s5bug commented:
What's required to push this PR to the finish line? I have a system similar to
```scala
trait T[X]
case object Foo extends T[Unit]

trait AsParam[L] {
  val tl: T[L]
}

def testParam[A](ap: AsParam[A]): Unit =
  ap.tl match {
    case Foo => println(summon[A =:= Unit])
    case _ => ()
  }
```
that I'd like to refactor into the equivalent of (the currently non-compiling on mainline Dotty)
```scala
trait AsMember {
  type L
  val tl: T[L]
}

def testMember(am: AsMember): Unit =
  am.tl match {
    case Foo => println(summon[am.L =:= Unit])
    case _ => ()
  }
```
so that methods can return `AsMember` instead of `AsParam[?]`, which requires such like
```scala
val asParamQ: AsParam[?] = someFunction
asParamQ match {
  case apt: AsParam[t] => apt.tl match {
    case Foo => println(summon[t =:= Unit])
    case _ => ()
  }
}
```
to consume (which does happen to compile on mainline).

Linyxus commented:
Hi! Glad to see this PR can benefit real-world development. It is currently blocked by #14776. This is a bug discovered when developing this PR and the tests currently broken would work after fixing this issue.

Linyxus commented:
Just verified that the following snippet compiles on this branch. :)

```scala
trait T[X]
case object Foo extends T[Unit]

trait AsMember {
  type L
  val tl: T[L]
}

def testMember(am: AsMember): Unit =
  am.tl match {
    case Foo => println(summon[am.L =:= Unit])
    case _ => ()
  }
```

Linyxus commented:
test performance please

Linyxus commented:
test performance please

Linyxus commented:
Fixes #15868.

abgruszecki commented on compiler/src/dotty/tools/dotc/core/GadtConstraint.scala:
```diff
@@ -11,38 +11,95 @@ import collection.mutable
+  def isEquivalent(p: PathType, q: PathType): Boolean
+
+  /** Query the representative member of a singleton type. */
+  def reprOf(p: PathType): PathType | Null
```
This function is part of the union-find related API, right? Could you list all the methods which are strictly only for union-find?

abgruszecki replied:
Also, could you explain why exactly we need `reprOf`? If we equate multiple type variables, we don't need a representative for them.

Linyxus replied:
Fixed! I looked at it and found that we do not actually need it. I removed this method and renamed various things to make the API related to singleton aliasing more clear.

Linyxus replied:
The only two APIs related to singleton aliasing (or the union-find) are: `recordPathAliasing` which records the alias between two paths and `isAliasingPath` which asks whether two paths are equivalent via path aliasing.

Linyxus replied:
The internal mapping is now named `pathAliasingMapping`.

abgruszecki commented on compiler/src/dotty/tools/dotc/core/GadtConstraint.scala:
```diff
@@ -11,38 +11,95 @@ import collection.mutable
+  def reprOf(p: PathType): PathType | Null
+
+  /** Scrutinee path of the current pattern matching. */
+  def scrutineePath: TermRef | Null
```
Could you explain why do we need `scrutineePath`, and why we need to reset it?

Linyxus replied:
I added detailed documentation in the docstring of `constrainTypeMembers` ([commit1](https://github.com/lampepfl/dotty/pull/14754/commits/e043fdd14979b0503794a378e94b45c8cf0c3e9c), [commit2](https://github.com/lampepfl/dotty/pull/14754/commits/7d5046ff97d807e7d5e1d28e44250c2b6fbc5fb7)). Does it look good?

abgruszecki commented on compiler/src/dotty/tools/dotc/core/GadtConstraint.scala:
```diff
@@ -63,31 +120,281 @@ final class ProperGadtConstraint private(
+    wasConstrained = false,
+    myScrutineePath = null,
+    myUnionFind = SimpleIdentityMap.empty,
+    myPatternSkolem = null,
```
Why is it necessary to keep this reference here?

Linyxus replied:
Yes we have to keep track of the skolem created as a placeholder for the pattern path, so that when the real pattern path is available we can substitute the placeholder with the actual path. But maybe we should move this reference to somewhere else?

Linyxus commented:
Hey Alex! @abgruszecki I have finished one round of PR polishing. I have done the following things:

- I have refactored the interface of path aliasing constraints. It is now more modular.
- I fixed the `typeMemberTouched` bug we discussed before, companied with a test. I also refactored the recursive calls to `constrainPatternType` by moving the logic into a `recur` function. So now we do not have `typeMemberTouched` argument at toplevel.
- Fixed one bug related to pattern path.
- Added tons of documentation.

There are one thing we discussed before but not in the review comments yet, so I am not sure whether I should do it: refactoring the states related to pattern and scrutinee paths into another class in `Context`, instead of keeping them in `GadtConstraint`. I am not sure what we should do here because the states for recording the scrutinee and the pattern paths are closely related to GADT reasoning and the way we implement path-dependent GADT. But maybe refactor them into another type will make things more clear and modular?

ckipp01 commented:
Hey @Linyxus are you willing to give this a rebase? If so, I think we can try and get this reviewed.

Linyxus commented:
Similarly, we will wait for a refactorization of GADT (https://github.com/lampepfl/dotty/pull/15928) before proceeding this feature. I will close this for now and go back to it after the refactorization is done.

LPTK commented:
https://github.com/lampepfl/dotty/pull/15928 was closed. Is there any future for the changes in this PR?

SethTisue commented:
@dwijnand maybe you could add a short summary here of where things stand?
